Return of abel.transform() class

stggh opened the pull request:
As per #147 I have returned `abel.transform.transform()` to a class.

>The previous syntax of `abel.transform(IM)['transform']`
   has been replaced with `abel.transform(IM).transform`

Main changes (from my memory):
1. wrapped the old `transform()` in a class, and restuctured.
2. updated `abel/examples` to use the class attribute `.transform`.
3. updated `abel/tests` as necessary.
4. updated the `abel/doc` files. NB to be further improved to better explain class structure.
    On my Mac the docs build using `make html`, but `make livehtml` failed.

This is a, I believe, fully functional first pass. Please left me know your thoughts.

```
In [1]: import abel
In [2]: IM = abel.tools.analytical.sample_image()
In [3]: IMobj = abel.transform(IM, direction="forward", method="hansenlaw")
In [4]: IMobj.
IMobj.IM         IMobj.direction  IMobj.method     IMobj.transform
```

DanHickstein commented:
@stggh, thanks for this huge effort! I will have a look now and see if I can catch any typos...

DanHickstein commented on doc/transform_methods/basex.rst:
```diff
@@ -38,9 +38,9 @@ According to Dribinski et al. BASEX has several advantages:
 -------------
 
-The recommended way to complete the inverse Abel transform using the BASEX algorithm for a full image is to use the :func:`abel.transform` function: ::
+The recommended way to complete the inverse Abel transform using the BASEX algorithm for a full image is to use the :func:`abel.transform` class: ::
```
`:func:` should be `:class:`

DanHickstein commented:
In order to get the docs to build the new class properly (especially to include the __init__ function), you'll want to change the following text in `doc/abel.rst`:
```
abel.transform module
---------------------

.. automodule:: abel.transform
    :members:
    :show-inheritance:
```
becomes:
```
abel.transform module
---------------------

.. autoclass:: abel.transform
    :members:
    :special-members;
    :show-inheritance:
```

btw, `make livehtml` is working on my Mac. Did you `pip install sphinx-autobuild`?

DanHickstein commented on README.rst:
```diff
@@ -76,10 +76,10 @@ Using PyAbel can be simple. The following Python code imports the PyAbel package
+	inverse_abel = abel.transform(forward_abel, direction='inverse', method='three_point').transform
 
-Note: the ``abel.transform()`` function returns a Python ``dict`` object, where the 2D Abel transform is accessed through the ``'transform'`` key.
+Note: the ``abel.transform()`` class returns a Python ``class`` object, where the 2D Abel transform is accessed through the ``.transform`` attribute.
```
I'm out of my element here, but I think that the language is that "abel.transform() returns an object (an instance of the abel.Transform class)

DhrubajyotiDas replied:
[Relevant Python doc link](https://docs.python.org/2/tutorial/classes.html#class-objects)

I agree with @DanHickstein.

DanHickstein commented:
In Python, class names should always be in CapWords (per [this section of PEP8](https://www.python.org/dev/peps/pep-0008/#class-names)). So, `abel.transform.transform` should become `abel.transform.Transform`. (of course, due to the `from .transform import transform` line in the main `__init__.py` file, we can just refer to it as `abel.Transform`. 

Sorry about this one. Probably annoying to change, but I think that the capitalization of classes is a fairly rigorous standard.

DanHickstein commented on abel/hansenlaw.py:
```diff
@@ -47,15 +47,15 @@ def hansenlaw_transform(IM, dr=1, direction="inverse"):
 
     Evaluation follows Eqs. (15 or 17), using (16a), (16b), and (16c or 18) of the Hansen 
-    and Law paper. For the full image transform, use ``abel.transform``.
+    and Law paper. For the full image transform, use the class ``abel.transform``.
```
change to ``:class:`abel.Transform``` in order to get a hyperlink in the docs.

DanHickstein commented on doc/transform_methods/hansenlaw.rst:
```diff
@@ -111,10 +111,10 @@ How to use it
 
 To complete the forward or inverse transform of a full image with the 
-``hansenlaw method``, simply use the :func:`abel.transform` function: ::
+``hansenlaw method``, simply use the :class:`abel.transform`: ::
```
I think that you should take out the word "the" (or add the word "class").

DanHickstein commented:
Okay, I looked through everything. I think that's all I got. If @DhrubajyotiDas and/or @ldes89150 have a chance to glance at the new `transform.py` that would be helpful.

ldes89150 commented on abel/transform.py:
I would prefer not to keep the original image if I want to use pickle to store the transformed results.

DanHickstein commented on abel/transform.py:
Oh, interesting. I was thinking that it would be a good idea to keep the original image. This would allow for the addition of functions that, perhaps, transform back and see how well the original image is reproduced. This could provide some error estimate for the transform. 

It seems like those interested in saving disk space could just save the `.transform`, but I suppose that throws away the meta information, such as the transform method used.

DhrubajyotiDas commented:
Thanks for taking the time to work on this massive PR, @stggh!

stggh commented:
No worries, I caused the rewrite in the first place. I am open to all suggestions on how to improve this code, or whether it is in fact a worthwhile improvement over the old `transform/dict` code. Thanks, for all the corrections and suggestions so far.

The pickle aspect, I had not considered. If `Transform` is just a wrapper class to return the calculations, its current role, then there is no problem, `self.IM` may be dropped. I privatized the class methods so that the transform results, from `IMobj.<tab>` were not lost in a maze of function names. Alternatively, if `Transform` becomes a true class container, then it might hold the original image, and provide for additional image operations. It is all a matter of user preference. For the moment we may stick with the class wrapper results only approach.

I will make further changes once I arrive at the ANU.

DanHickstein commented:
Yes, at this point I think that there is little utility in storing the original image in the class. It just takes up space. In the future, if we want to allow the class to perform functions that need the original image (i.e., to compare to the original image and find the error), then we can re-assess the need for the class to keep the original image.

stggh commented:
I have made the changes that you kindly identified above.

Notes:
 1. I 'cheated' and simply `del self._IM` following the `abel_transform_image()` call. In principle, I could pass IM to each method, but this became messy if `IM` needs to be returned, e.g. for `recast_as_float64`. Let me know if this change is sufficient, or there is a better way.

 2. @DanHickstein sphinx guru - great information about this complex, but looks great, doc system.

 3.  Note the `<tabs>` were added by you during your (amazing) doc edits. I have removed all traces.

 4. `abel.transform` is now `abel.Transform`, tks for setting me straight.

 5. Updated all examples and unit tests.

 6. `make livehtml` has, in the past, run correctly on my Mac, so there should be no sphinx components missing.   However, my Linux system can't find the napoleon module, which appears to be a bug present in Ubuntu.

I will run further tests today, and check through the docs.

DanHickstein commented:
Great work @stggh!

That "cheat" sounds fine. If it's too difficult to remove the original image, then we can just keep it. I suspect that most people will save the 2D array anyway. 

Whoops! Sorry about those tabs. :)

There are a few warnings when building the docs. They were probably not all caused by you, but can you take a look?

```
docstring of abel.transform.Transform.__init__:69: ERROR: Unexpected indentation.
docstring of abel.transform.Transform.__init__:72: ERROR: Unexpected indentation.
docstring of abel.transform.Transform.__init__:74: WARNING: Definition list ends without a blank line; unexpected unindent.
/Users/ddh/Documents/PyAbel-master/abel/tools/symmetry.py:docstring of abel.tools.symmetry.get_image_quadrants:24: ERROR: Unexpected indentation.
/Users/ddh/Documents/PyAbel-master/abel/tools/symmetry.py:docstring of abel.tools.symmetry.get_image_quadrants:27: ERROR: Unexpected indentation.
```

The livehtml was also not working for me for a while. I forget what the solution was. You have tried `make clean` right?

DanHickstein commented:
...and while you are looking at the docs, can you fix the indentation of this section of the `abel.transform.Transform.__init__` docstring:

![image](https://cloud.githubusercontent.com/assets/1107796/13830806/ce3dc3dc-eb94-11e5-868d-8ed77950d864.png)

stggh commented:
I did not identify anything obvious with `abel.transform.Transform.__init__` doc string errors, identified above. The `symmetrizing` doc string is repaired (my fault for attempting to removed long lines).

I did forget the `make clean`. The Mac is at home, so I cannot test further.

The napoleon Ubuntu issue is weird.  I have repeated all the `pip install` commands 
```
make html
sphinx-build -b html -d _build/doctrees   . _build/html
Making output directory...
Running Sphinx v1.2.3

Extension error:
Could not import extension sphinx.ext.napoleon (exception: No module named napoleon)
```

It says here in this readthedocs [forum](https://github.com/rtfd/readthedocs.org/issues/1164)
>Seems this was fixed with the Sphinx 1.3.1 update. 

However, my sphinx version appears to be 1.2.3?

DanHickstein commented:
Can you update Sphinx from 1.2 to 1.3?

stggh commented:
Yes! Removal of the ubuntu sphinx-common (and associated packages) and then `pip install -U Sphinx` appears to have worked. Thanks!

stggh commented:
There is a case for keeping `self._IM` as it is the `centered` odd-size (width) image, which might speed up subsequent calls to `abel.Transform()`. Perhaps, (yet) another parameter is needed in the class instance function?

DanHickstein commented:
It's fine with me to keep `self._IM` around. @ldes89150 will just need a bigger pickle-jar :)

ldes89150 commented:
@stggh thanks for fixing the doc string :D

It's also fine for me to keep `self._IM`. I can save `self.transform` only.

DanHickstein commented:
Awesome! Are we ready to merge this then?

stggh commented:
I have put back  `self.IM` as it is the image used in the transform, which may differ from `IM` (centered, odd-width), removed the `del` and updated the `Transform` doc string.

If this change is ok, then we are ready to merge.
